app/consumer/genesis.go

```diff
@@ -15,7 +28,130 @@ import (
 // object provided to it during init.
 type GenesisState map[string]json.RawMessage
 
+type (
+	// Transformation callback for a consumer genesis 'section' exported from a
+	// specific version of a provider
+	TransformationCallback func([]byte, client.Context) (json.RawMessage, error)
+
+	// TransformationMap defines the mapping from a version to a transformation callback
+	TransformationMap map[string]TransformationCallback
+)
+
+// Migration of consumer genesis content as it is exported from a provider version v2
+// (ICS provider module version) to a format readable by current consumer implementation.
+func migrateFromV2(jsonRaw []byte, ctx client.Context) (json.RawMessage, error) {
+	// v2 uses deprecated fields of GenesisState type
+	oldConsumerGenesis := consumerTypes.GenesisState{}
+	err := ctx.Codec.UnmarshalJSON(jsonRaw, &oldConsumerGenesis)
+	if err != nil {
+		return nil, fmt.Errorf("reading consumer genesis data failed: %s", err)
+	}
+
+	// some sanity checks for v2 transformation
+	if len(oldConsumerGenesis.Provider.InitialValSet) > 0 {
+		return nil, fmt.Errorf("invalid source version. Unexpected element 'provider.initial_val_set'")
+	}
+
+	if oldConsumerGenesis.Provider.ClientState != nil {
+		return nil, fmt.Errorf("invalid source version. Unexpected element 'provider.client_state'")
+	}
+
+	if oldConsumerGenesis.Provider.ConsensusState != nil {
+		return nil, fmt.Errorf("invalid source version. Unexpected element 'provider.consensus_state'")
+	}
+
+	// Version 2 of provider genesis data fills up deprecated fields
+	// ProviderClientState, ConsensusState and InitialValSet
+	newGenesis := types.ConsumerGenesisState{
+		Params: oldConsumerGenesis.Params,
+		Provider: types.ProviderInfo{
+			ClientState:    oldConsumerGenesis.ProviderClientState,
+			ConsensusState: oldConsumerGenesis.ProviderConsensusState,
+			InitialValSet:  oldConsumerGenesis.InitialValSet,
+		},
+	}
+
+	newJson, err := ctx.Codec.MarshalJSON(&newGenesis)
+	if err != nil {
+		return nil, fmt.Errorf("failed marshalling data to new type: %s", err)
+	}
+	return newJson, nil
+}
+
+// Mapping of provider module version to related transformation function
+var transformationMap = TransformationMap{
```

Comment:
I think we can remove this and simply comment what the CLI `transform` does.

The `transform` should only do the following:
* transform consumer genesis output coming from providers running ICSv1/2/3 into the latest consumer genesis version